File: mojo/public/python/mojo_application/application_runner.py
# Copyright 2015 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

"""Helper for running Mojo applications in Python."""

from mojo_application.application_impl import ApplicationImpl

import mojo_system

def RunMojoApplication(application_delegate, app_request_handle):
  loop = mojo_system.RunLoop()

  application = ApplicationImpl(application_delegate, app_request_handle)
  application.manager.AddOnErrorCallback(loop.Quit)

  loop.Run()

File: mojo/public/cpp/application/lib/application_runner.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/public/cpp/application/application_runner.h"

#include "mojo/public/cpp/application/application_delegate.h"
#include "mojo/public/cpp/application/application_impl.h"
#include "mojo/public/cpp/environment/environment.h"
#include "mojo/public/cpp/utility/run_loop.h"

namespace mojo {

// static
void ApplicationImpl::Terminate() {
  RunLoop::current()->Quit();
}

ApplicationRunner::ApplicationRunner(ApplicationDelegate* delegate)
    : delegate_(delegate) {
}
ApplicationRunner::~ApplicationRunner() {
  assert(!delegate_);
}

MojoResult ApplicationRunner::Run(MojoHandle app_request_handle) {
  Environment env;
  {
    RunLoop loop;
    ApplicationImpl app(delegate_, MakeRequest<Application>(MakeScopedHandle(
                                       MessagePipeHandle(app_request_handle))));
    loop.Run();
  }

  delete delegate_;
  delegate_ = nullptr;
  return MOJO_RESULT_OK;
}

}  // namespace mojo
